exporter/otlpexporter/arrow.go

```diff
@@ -0,0 +1,419 @@
+	case stream := <-e.arrow.ready.readyChannel():
+		return stream, nil
+	}
+}
+
+// shutdown returns when all Arrow-associated goroutines have returned.
+func (ae *arrowExporter) shutdown(ctx context.Context) error {
+	ae.cancel()
+	ae.wg.Wait()
+	return nil
+}
+
+// setBatchChannel places a waiting consumer's batchID into the waiters map, where
+// the stream reader may find it.
+func (stream *arrowStream) setBatchChannel(batchID string, errCh chan error) {
+	stream.lock.Lock()
+	defer stream.lock.Unlock()
+
+	stream.waiters[batchID] = errCh
+}
+
+// writeStream repeatedly places this stream into the next-available queue, then
+// performs a blocking send().  This returns when the data is in the write buffer,
+// the caller waiting on its error channel.
+func (ae *arrowExporter) writeStream(ctx context.Context, stream *arrowStream, ww *sync.WaitGroup) {
+	defer ww.Done()
+	defer stream.cancel()
+
+writeLoop:
+	for {
+		// Note: this can't block b/c stream has capacity &
+		// individual streams shut down synchronously.
+		ae.ready.setReady(stream)
+
+		// this can block, and if the context is canceled we
+		// wait for the reader to find this stream.
+		var wri writeItem
+		select {
+		case wri = <-stream.toWrite:
+		case <-ctx.Done():
+			// Because we did not <-stream.toWrite, there is a potential
+			// sender race happening, which is handled below.
+			break writeLoop
+		}
+		// Note: For the two return statements below there is no potential
+		// sender race because the stream is not available, as indicated by
+		// the successful <-stream.toWrite.
+
+		batch, err := stream.encode(wri.records)
+		if err != nil {
+			// TODO: Is this not permanent?  Another
+			// sequence of data might not produce it.
+			//
+			// This is some kind of internal error.
+			wri.errCh <- consumererror.NewPermanent(err)
+			ae.exporter.settings.Logger.Error("arrow encode", zap.Error(err))
+			return
+		}
+
+		stream.setBatchChannel(batch.BatchId, wri.errCh)
+
+		if err := stream.client.Send(batch); err != nil {
+			// The error will be sent to errCh during cleanup for this stream.
+			ae.exporter.settings.Logger.Error("arrow send", zap.Error(err))
+			return
+		}
+	}
+
+	// Remove this from the ready channel.
+	ae.ready.removeReady(stream)
+}
+
+// readStream repeatedly reads a batch status and releases the consumers waiting for
+// a response.
+func (ae *arrowExporter) readStream(ctx context.Context, stream *arrowStream) error {
+	defer stream.cancel()
+
+	for {
+		resp, err := stream.client.Recv()
+		if err != nil {
+			return err
+		}
+
+		ae.processBatchStatus(stream, resp.Statuses)
+	}
+}
+
+// processBatchStatus processes a single response from the server.
+func (ae *arrowExporter) processBatchStatus(stream *arrowStream, statuses []*arrowpb.StatusMessage) {
+	stream.lock.Lock()
+	defer stream.lock.Unlock()
+
+	for _, status := range statuses {
+		ch, ok := stream.waiters[status.BatchId]
+		if !ok {
+
+			continue
+		}
+
+		ch <- nil
```

Comment:
Is there a guarantee that the channel will always be able to accept this message without waiting? If not, then I suggest to reorganize the code in order to minimize the section where the lock is active (i.e. move the lock section inside the loop, move the delete into this smaller section, and then interact with the channel outside the lock section).

Reply:
Yes, there's a guarantee because the channel was created with size 1. However, I like your suggestion that the channel operations be moved outside the lock. Will do.

Reply:
I've added a `getSenderChannels()` operation that takes the lock and returns a slice of channels.